apps/src/applab/Exporter.js

```diff
@@ -337,12 +379,183 @@ export default {
-    return this.exportAppToZip(appName, code, levelHtml)
+  async exportApp(appName, code, levelHtml, suppliedExpoOpts) {
+    const expoOpts = suppliedExpoOpts || {};
+    if (expoOpts.mode === 'publish') {
+      const session = await this.publishToExpo(appName, code, levelHtml);
+      const saveResult = await session.saveAsync();
+      const expoURL = `exp://expo.io/@snack/${saveResult.id}`;
+      return expoURL;
+    }
+    return this.exportAppToZip(appName, code, levelHtml, expoOpts.mode === 'zip')
       .then(function (zip) {
         zip.generateAsync({type:"blob"}).then(function (blob) {
           saveAs(blob, appName + ".zip");
         });
       });
+  },
+
+  createPackageFilesFromZip(zip, appName) {
+    const moduleList = [];
+    zip.folder(appName + "/assets").forEach((fileName, file) => {
+        if (!file.dir) {
+          moduleList.push({ fileName });
+        }
+    });
+    const entries = moduleList.map(module => exportExpoPackagedFilesEntryEjs({ module }));
+    return exportExpoPackagedFilesEjs({ entries });
+  },
+
+  createPackageFilesFromExpoFiles(files) {
+    const moduleList = [];
+    const assetPrefix = "assets/";
+    const assetPrefixLength = assetPrefix.length;
+    for (const fileName in files) {
+      if (fileName.indexOf(assetPrefix) !== 0) {
+        continue;
+      }
+      const relativePath = fileName.substring(assetPrefixLength);
+      moduleList.push({ fileName: relativePath });
+    }
+    const entries = moduleList.map(module => exportExpoPackagedFilesEntryEjs({ module }));
+    return exportExpoPackagedFilesEjs({ entries });
+  },
+
+  publishToExpo(appName, code, levelHtml) {
+    const appJson = exportExpoAppJsonEjs({
+      appName: appName,
+      projectId: project.getCurrentId()
+    });
+    const appOptionsJs = getAppOptionsFile();
+    const { css, outerHTML } = transformLevelHtml(levelHtml);
+    const html = exportExpoIndexEjs({
+      htmlBody: outerHTML,
+      commonLocalePath: "https://studio.code.org/blockly/js/en_us/common_locale.js",
+      applabLocalePath: "https://studio.code.org/blockly/js/en_us/applab_locale.js",
+      appOptionsPath: "appOptions.j",
+      applabApiPath: "https://studio.code.org/blockly/js/applab-api.js",
+      jQueryPath: "https://code.jquery.com/jquery-1.12.1.min.js",
+      commonCssPath: "https://studio.code.org/blockly/css/common.css",
+      applabCssPath: "https://studio.code.org/blockly/css/applab.css",
+    });
+
+    const appAssets = dashboard.assets.listStore.list().map(asset => ({
+      url: assetPrefix.fixPath(asset.filename),
+      rootRelativePath: asset.filename,
+      dataType: 'binary',
+      filename: asset.filename,
+    }));
+
+    const files = {
+      'App.js': { contents: exportExpoAppJs, type: 'CODE'},
+      'CustomAsset.js': { contents: exportExpoCustomAssetJs, type: 'CODE'},
+      'app.json': { contents: appJson, type: 'CODE'},
+    };
+
+    return new Promise(async (resolve, reject) => {
```

Comment:
Okay, learning `async/await` syntax so help me out. My understanding is that `async` functions return a Promise, and implicitly `resolve()` their return value, and `reject` with thrown exceptions.

Here it looks like you only `resolve` in your return statement, and don't use `reject` at all.  Does that mean you could express this:
```js
return new Promise(async (resolve, reject) => {
  // Long-running work...
  return resolve(session);
});
```
As this?
```js
return (async () => {
  // Long-running work...
  return session;
})();
```
Or, since `publishToExpo` is returning a Promise anyway, I think you could just make _it_ async and skip the IIFE in the middle.
```js
async publishToExpo(appName, code, levelHtml) {
  // Long-running work...
  return session;
}
```

Reply:
yes, I think this last option is the best. Seems like you are getting it!